### PR TITLE
Excess functions

### PR DESCRIPTION
**Pull request by cicr99**
The purpose of this pass is to find functions like the following and remove them as a way to optimize the code:
```
function foo(uint a) returns (uint) {
  return a;
}
```
If these functions are not externally visible, it is possible to remove them and replace the calls to that function with the expression it returns, tidying up the references of course.
This pass also recognizes functions of this type that return more than one argument in the form of a `TupleExpression`, even if the identifiers are not in the same order as the function parameters.

### CODE REVIEW

File: tests/behaviour/expectations/behaviour.ts [hunk 1]
```diff
@@ -3792,6 +3792,9 @@ export const expectations = flatten(
           Expect.Simple('test', [], []),
           Expect.Simple('returnTest', [], ['12', '0']),
         ]),
+        File.Simple('removeIdentityFunctions', [
+          Expect.Simple('f', ['10', '0'], ['3', '20', '0', '5', '0', '13', '0']),
+        ]),
```
Comment (rodrigo-pino) [inline tests/behaviour/expectations/behaviour.ts:3797]: Is there a way of testing that function `h` and `g` got pruned correctly. Like calling them and catching the error because they do not exist

**Review by rodrigo-pino — APPROVED** [review]
It looks good to me

File: tests/behaviour/expectations/behaviour.ts (hunk 1 above)
Comment (cicr99) [reply tests/behaviour/expectations/behaviour.ts:3797]: The thing is for those functions to be pruned they most not be externally visible, so even if there was no pruning at all, calling them should always throw an error. In other words, I think that trying to call those functions won't work as a way of testing this pass.

Comment (rodrigo-pino) [reply tests/behaviour/expectations/behaviour.ts:3797]: Right, that's true

File: src/passes/functionPruner/index.ts [hunk 2]
```diff
@@ -3,13 +3,7 @@ import { ASTMapper } from '../../ast/mapper';
 import { CallGraphBuilder } from './callGraph';
 import { FunctionRemover } from './functionRemover';
 
-export class UnreachableFunctionPruner extends ASTMapper {
-  // Function to add passes that should have been run before this pass
```
Comment (GuyMcComb) [inline src/passes/functionPruner/index.ts:7]: Why was this deleted?

File: src/passes/functionPruner/index.ts [hunk 3]
```diff
@@ -3,13 +3,7 @@ import { ASTMapper } from '../../ast/mapper';
...
-  addInitialPassPrerequisites(): void {
```
Comment (GuyMcComb) [inline src/passes/functionPruner/index.ts:8]: Why was this removed?

**Comment by GuyMcComb** [comment]
Sem tests running

File: src/passes/functionPruner/index.ts (hunk 3 above)
Comment (cicr99) [reply src/passes/functionPruner/index.ts:8]: Oh I didn't realize it got deleted. It must have been when I had to fix some conflicts. I'll add it

**Comment by GuyMcComb** [comment]
Sem test pass. Just waiting for deleted block to be added back

File: src/passes/identityFunctionRemover.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  Identifier,
+  Return,
+  TupleExpression,
+} from 'solc-typed-ast';
+import { AST } from '../ast/ast';
+import { ASTMapper } from '../ast/mapper';
+import { printNode } from '../utils/astPrinter';
+import { isExternallyVisible, toSingleExpression } from '../utils/utils';
+
+export class IdentityFunctionRemover extends ASTMapper {
```
Comment (GuyMcComb) [inline src/passes/identityFunctionRemover.ts:15]: Looks good, just needs some documentation describing this pass.

File: src/passes/functionPruner/index.ts (hunk 3 above)
Comment (GuyMcComb) [reply src/passes/functionPruner/index.ts:8]: Might be a good opportunity to add some documentation to this pass as well.

**Review by GuyMcComb — CHANGES_REQUESTED** [review]

**Comment by GuyMcComb** [comment]
Sem Tests passed with new commits

**Review by GuyMcComb — APPROVED** [review]
